src/main/resources/crafter/studio/upgrade/3.1.x/3.1.0.29/site/config-list.xslt

```diff
@@ -0,0 +1,116 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright (C) 2007-2019 Crafter Software Corporation. All Rights Reserved.
+  ~
+  ~ This program is free software: you can redistribute it and/or modify
+  ~ it under the terms of the GNU General Public License as published by
+  ~ the Free Software Foundation, either version 3 of the License, or
+  ~ (at your option) any later version.
+  ~
+  ~ This program is distributed in the hope that it will be useful,
+  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  ~ GNU General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU General Public License
+  ~ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  -->
+
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="2.0">
+
+    <!-- to keep the right formatting -->
+    <xsl:output method="xml" indent="yes"/>
+    <xsl:strip-space elements="*"/>
+
+    <!-- copy all elements -->
+    <xsl:template match="node() | @*">
+        <xsl:copy>
+            <xsl:apply-templates select="node() | @*"/>
+        </xsl:copy>
+    </xsl:template>
+
+    <!-- insert line breaks before and after top level comments -->
+    <xsl:template match="/comment()">
+        <xsl:text>&#10;</xsl:text>
+        <xsl:copy-of select="."/>
+        <xsl:text>&#10;</xsl:text>
+    </xsl:template>
+
+    <xsl:template name="search-and-replace">
```

Comment:
If this is a regular text operation we have an `FindAndReplaceUpgradeOperation` that could be used to avoid repeating this in all the XSLT files and run it once for all config files that need it

Reply:
This XSLT is a complex transformation and `search-and-replace` is part of it. In general use case,  XSLT can be replaced with string + regex operations. Then we don't need XSLT upgrader at all, no?
Personally, XSLT gives more flexibility while writing it. I can produce and test XSLT using online tester tools and when I am happy with the result I can add the final version to upgrade pipeline. I find this process faster and more comfortable than iterating build-start-stop studio to test 'FindAndReplaceUpgradeOperation'

Reply:
find . -name 'xyz.xml'| xargs perl -pi -e "s/before/after/g;" is just as easy to do without build/start/stop.

Having said that, it's fine to do things in XSLT if that works for you. XSLT won't work for everything, however, so we do need our FindAndReplaceUpgradeOperation (and it's faster).

Reply:
Yeah, I was just suggesting to do XSLTs just to add the new nodes and then a general FindAndReplace to simplify the templates.

Reply:
if you review the whole XSLT you will see it adds a new node and replaces the content at same time